Revamp quick-tour (#1469)
ballerina-platform/ballerina-www: website/two-column-pages/docs/learn/quick-tour.md

// A system module containing protocol access constructs
// Module objects referenced with 'http:' in code
import ballerina/http;
import ballerina/io;

# A service is a network-accessible API
# Advertised on '/hello', port comes from listener endpoint
service hello on new http:Listener(9090) {

    # A resource is an invokable API method
    # Accessible at '/hello/sayHello

>     # 'caller' is the client invoking this resource 

Reviewer: ```suggestion
    # 'caller' is the client invoking this resource.
```

There are two kinds of endpoints in Ballerina, inbound (or ingress) and outbound (or egress) endpoints, a client object is an outbound endpoint, which we would use to send messages to a network service.

> Having said that, let's see how you can make hello_service a bit more interesting by invoking an external endpoint. In this case, sunrise-sunset API is used as the backend service. First, you need to created the client with the relevant endpoint URL as follows.

Reviewer: ```suggestion
Having said that, let's see how you can make the `hello_service` a bit more interesting by invoking an external endpoint. In this case, the `sunrise-sunset` API is used as the backend service. First, you need to create the client with the relevant endpoint URL as follows.
```

```ballerina 
http:Client weatherApi = new("https://samples.openweathermap.org");
```

As the next step, let's tweak the service a bit to get sunrise/sunset time details for London. 

1. Change the service name to `sunriseSunset` and resource name to `london`.
2. Update the resource signature to have `error?` as the return value.

> > **Note**: returning `error?` allows you to use `check` key word to avoid handling errors explicitly. This is only done to keep the code simple. But in real production code you may have to explicitly handle those errors. 

Reviewer: ```suggestion
> **Note**: Returning `error?` allows you to use the `check` keyword to avoid handling errors explicitly. This is only done to keep the code simple. However, in real production code, you may have to explicitly handle those errors. 
```

Following is the sample code with the aforementioned changes:

```ballerina

> # A service is a network-accessible API

Reviewer: ```suggestion
# A service is a network-accessible API.
```

# Advertised on '/sunriseSunset', port comes from listener endpoint
service sunriseSunset on new http:Listener(9090) {

    # A resource is an invokable API method

>     # Accessible at '/hello/sayHello

Reviewer: ```suggestion
    # accessible at '/hello/sayHello
```

>     # 'caller' is the client invoking this resource 

Reviewer: ```suggestion
    # 'caller' is the client invoking this resource.
```

    # + caller - Server Connector
    # + request - Request

>     # + return - error, if there is an issue

Reviewer: ```suggestion
    # + return - Error, if there is an issue
```

    resource function london(http:Caller caller, http:Request request) returns error? {
```

As the next step, add the below code line to do a Get request to the sunrise-sunset backend.

```ballerina
http:Response sunriseResp = check sunriseApi->get("/json?lat=51.5074&lng=0.1278");
```

As the next step, add the below code snippet to retrieve the payload and transform it to a simpler response payload.

```ballerina
json sunrisePayload = check sunriseResp.getJsonPayload();
// Creates the response payload
json resPayload = {
    city: "London",
    sunrise: check sunrisePayload.results.sunrise,
    suset: check sunrisePayload.results.sunset
};
```

> As the last step you need to change the response payload that is being passed to `respond` action with the new json payload as shown below.

Reviewer: ```suggestion
As the last step, you need to change the response payload that is being passed to the `respond` action with the new JSON payload as shown below.
```

```ballerina
var result = check caller->respond(resPayload);
```

> The complete source code should like similar to the following:

Reviewer: ```suggestion
The complete source code should look similar to the following:
```

```ballerina
// A system module containing protocol access constructs
// Module objects referenced with 'http:' in code
import ballerina/http;
import ballerina/io;

http:Client sunriseApi = new("http://api.sunrise-sunset.org");

> # A service is a network-accessible API

Reviewer: ```suggestion
# A service is a network-accessible API.
```

# Advertised on '/sunriseSunset', port comes from listener endpoint
service sunriseSunset on new http:Listener(9090) {

>     # A resource is an invokable API method

Reviewer: ```suggestion
    # A resource is an invokable API method.
```

    # Accessible at '/hello/sayHello
    # 'caller' is the client invoking this resource 

    # + caller - Server Connector
    # + request - Request

>     # + return - error, if there is an issue

Reviewer: ```suggestion
    # + return - Error, if there is an issue
```

    resource function london(http:Caller caller, http:Request request) returns error? {
        http:Response sunriseResp = check sunriseApi->get("/json?lat=51.5074&lng=0.1278");

        json sunrisePayload = check sunriseResp.getJsonPayload();

>         // Creates the response payload

Reviewer: ```suggestion
        // Creates the response payload.
```

        json resPayload = {
            city: "London",
            sunrise: check sunrisePayload.results.sunrise,
            suset: check sunrisePayload.results.sunset
        };

        error? result = caller->respond(resPayload);
        if (result is error) {
            io:println("Error in responding", result);
        }
    }
}
```

> Now, before we build the module let's change sample_service.bal to sunrise_sunset_service.bal.

Reviewer: ```suggestion
Now, before we build the module, let's change the `sample_service.bal` to `sunrise_sunset_service.bal`.
```

> All set. Once again let's build the module by running the below command.

Reviewer: ```suggestion
All set. Once again, let's build the module by running the below command.
```

```bash
$ ballerina build
```

> Now you can ran the service by running the following command. 

Reviewer: ```suggestion
Now, you can run the service by running the following command. 
```

```bash
$ ballerina run /<folder_path>/sample_service-executable.jar
```

As before, you should get the following output.

```bash
[ballerina/http] started HTTP/WS listener 0.0.0.0:9090
```

This means your service is up and running. You can invoke the new service by running the below cURL command.

```bash
$ curl -v http://localhost:9090/sunriseSunset/london
```

## Follow the Repo

<div class="cGitButtonContainer"><p data-button="iGitStarText">"Star"</p> <p data-button="iGitWatchText">"Watch"</p></div>

Star [GitHub repo](https://github.com/ballerina-platform/ballerina-lang) and show appreciation to Ballerina maintainers for their work. Watch the repo to keep track of Ballerina issues.

## What's Next

> Now that you have taken Ballerina around for a quick tour, you can explore Ballerina more.

Reviewer: ```suggestion
Now, that you have taken Ballerina around for a quick tour, you can explore Ballerina more.
```
